sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinInterfaceTest.scala

```diff
@@ -24,28 +23,160 @@ import org.scalatest.Inside
+  // The following code defines a package -iface-pkg- that defines a single interface Iface.
+  val ifacePkgName = Ref.PackageName.assertFromString("-iface-pkg-")
+  val ifacePkgId = Ref.PackageId.assertFromString("-iface-package-id-")
+  val ifaceParserParams = ParserParameters(
+    defaultPackageId = ifacePkgId,
+    // TODO: revert to the default version once it supports upgrades
+    languageVersion = LanguageVersion.Features.packageUpgrades,
+  )
+  val ifacePkg =
+    p"""metadata ( '$ifacePkgName' : '1.0.0' )
+            module Mod {
+              val mkParty : Text -> Party = \(t:Text) ->
+                case TEXT_TO_PARTY t of None -> ERROR @Party "none" | Some x -> x;
+
+              record @serializable MyViewType = { n : Int64 };
+              interface (this : Iface) = {
+                viewtype Mod:MyViewType;
+                choice @nonConsuming MyChoice (self) (u: Unit): Unit
+                  , controllers (Cons @Party [Mod:mkParty "Alice"] Nil @Party)
+                  , observers (Nil @Party)
+                  to upure @Unit ();
+              };
+            }
+          """ (ifaceParserParams)
+
+  // The following code defines a family of packages -implem-pkg- versions 1.0.0, 2.0.0, ... that define a
+  // template T that implements Iface. The view function for version 1 of the package returns 1, the view function
+  // of version 2 of the package returns 2, etc.
+  val implemPkgName = Ref.PackageName.assertFromString("-implem-pkg-")
+  def implemPkgVersion(pkgVersion: Int) =
+    Ref.PackageVersion.assertFromString(s"${pkgVersion}.0.0")
+  def implemPkgId(pkgVersion: Int) =
+    Ref.PackageId.assertFromString(s"-implem-pkg-id-$pkgVersion-")
+  def implemParserParams(pkgVersion: Int) = ParserParameters(
+    defaultPackageId = implemPkgId(pkgVersion),
+    // TODO: revert to the default version once it supports upgrades
+    languageVersion = LanguageVersion.Features.packageUpgrades,
+  )
+  def implemPkg(pkgVersion: Int) =
+    p"""metadata ( '$implemPkgName' : '${implemPkgVersion(pkgVersion)}' )
+            module Mod {
+              record @serializable T = { p: Party };
+              template (this: T) = {
+                precondition True;
+                signatories Cons @Party [Mod:T {p} this] (Nil @Party);
+                observers Nil @Party;
+                implements '$ifacePkgId':Mod:Iface { view = '$ifacePkgId':Mod:MyViewType { n = $pkgVersion }; };
+              };
+            }
+          """ (implemParserParams(pkgVersion))
+
+  // All three of -iface-package-id-, -implem-pkg-id-1-, and -implem-pkg-id-2- are made available to the interpreter.
+  val compiledPackages = PureCompiledPackages.assertBuild(
+    Map(
+      ifacePkgId -> ifacePkg,
+      implemPkgId(1) -> implemPkg(1),
+      implemPkgId(2) -> implemPkg(2),
+    ),
+    // TODO: revert to the default compiler config once it supports upgrades
+    Compiler.Config.Dev(LanguageMajorVersion.V2),
```

Comment:
can we factorize the version and the comment (lines 52, 82 and here) ?

Reply:
Done